source/volume_of_fluid/utilities.cc

```diff
@@ -0,0 +1,687 @@
+              }
+          }
+        //Simple cases
+        if (vol<=0.0)
+          {
+            return -0.5*norm1;
+          }
+
+        double m1 = nnormal[0];
+        double m2 = nnormal[1];
+        double m3 = nnormal[2];
+        double m12 = m1+m2;
+
+        // Case 1
+        double v1=0.0;
+        if (6*m2*m3>eps)
+          {
+            v1 = m1*m1/(eps);
+          }
+        else
+          {
+            v1 = m1*m1/(6*m2*m3);
+          }
+        if (vol<v1)
+          {
+            return -0.5+std::pow(6*mprod*vol, 1./3.);
+          }
+
+        // Case 2
+        double v2 = v1 + 0.5*(m2-m1)/m3;
+        if (vol<v2)
+          {
+            return 0.5*(-1+m1+sqrt(m1*m1+8*m2*m3*(vol-v1)));
+          }
+
+        // Case 3
+        double v3 = 0.0;
+        if (m12<m3)
+          {
+            v3 += m3*m3*(3*m12-m3)/(6.0*mprod);
+            v3 += m1*m1*(m1-3*m3)/(6*mprod);
+            v3 += m2*m2*(m2-3*m3)/(6*mprod);
+          }
+        else
+          {
+            v3 = 0.5*m12/m3;
+          }
+        if (vol<v3)
+          {
+            // Solve appropriate cubic
+            double a2 = -3.0*m12;
+            double a1 = 3.0*(m1*m1+m2*m2);
+            double a0 = 6*mprod*vol-m1*m1*m1-m2*m2*m2;
+            double np0 = a2*a2/9.0-a1/3.0;
+            double q0 = (a1*a2-3.0*a0)/6.0-a2*a2*a2/27.0;
+            double theta = acos(q0/sqrt(np0*np0*np0))/3.0;
+            return sqrt(np0)*(sqrt(3.0)*sin(theta)-cos(theta))-a2/3.0;
+          }
+
+        // Case 4
+        if (m3<m12)
+          {
+            // Solve appropriate cubic
+            double a2 = -1.5;
+            double a1 = 1.5*(m1*m1+m2*m2+m3*m3);
+            double a0 = 6*mprod*vol-m1*m1*m1-m2*m2*m2-m3*m3*m3;
+            double np0 = a2*a2/9.0-a1/3.0;
+            double q0 = (a1*a2-3.0*a0)/6.0-a2*a2*a2/27.0;
+            double theta = acos(q0/sqrt(np0*np0*np0))/3.0;
+            return sqrt(np0)*(sqrt(3.0)*sin(theta)-cos(theta))-a2/3.0;
+          }
+
+        return -0.5+m1*vol+0.5*m12;
+      }
+
+      template<>
+      void xFEM_Heaviside<2>(const int degree,
+                             const Tensor<1, 2> normal,
+                             const double d,
+                             const std::vector<Point<2>> &points,
+                             std::vector<double> &values)
+      {
+        const int basis_count=4;
+        std::vector<double> coeffs(basis_count);
+
+        const double n_xp = fabs(normal[0]), n_yp = fabs(normal[1]);
+        const double sign_n_x = (((normal[0]) > 0) - ((normal[0]) < 0)),
+                     sign_n_y = (((normal[1]) > 0) - ((normal[1]) < 0));
+
+        const double norm1 = n_xp + n_yp;
+        const double triangle_break = 0.5L*fabs(n_xp-n_yp);
+
+        const int max_degree = 1;
+
+        AssertThrow(degree<=max_degree,
+                    ExcMessage("Cannot generate xFEM polynomials are only functional for degrees<2."));
+
+        // Values calculated using sympy
```

Comment:
The algorithm below needs some explanation in this comment. Keep the reference to sympy, but also explain what you are doing.

Reply:
Clarification added